test/test_device_radix_sort.cu

```diff
@@ -626,73 +628,218 @@ template <typename KeyT>
 void InitializeKeyBits(
     GenMode         gen_mode,
     KeyT            *h_keys,
-    int             num_items,
+    size_t          num_items,
     int             /*entropy_reduction*/)
 {
-    for (int i = 0; i < num_items; ++i)
+    for (size_t i = 0; i < num_items; ++i)
         InitValue(gen_mode, h_keys[i], i);
 }
 
 
+/** Initialize the reference array monotonically. */
+template <typename KeyT>
+void InitializeKeysSorted(
+    KeyT *h_keys,
+    size_t num_items)
+{
+    using TraitsT = cub::Traits<KeyT>;
+    using UnsignedBits = typename TraitsT::UnsignedBits;
+
+    // Numbers to generate random runs.
+    UnsignedBits max_inc = 1 << (sizeof(UnsignedBits) < 4 ? 3 :
+                                 (sizeof(UnsignedBits) < 8 ? 14 : 24));
+    UnsignedBits max_bits = ~0;
+    size_t max_run = std::max(num_items / size_t(max_bits / max_inc), size_t(1 << 14));
+
+    UnsignedBits *h_key_bits = (UnsignedBits *)h_keys;
+    size_t i = 0;
+    // Start with the minimum twiddled key.
+    UnsignedBits twiddled_key = 0;
+    while (i < num_items)
+    {
+        // Generate random increment (avoid overflow).
+        UnsignedBits inc_bits = 0;
+        RandomBits(inc_bits);
+        // twiddled_key < bits_max at this point.
+        UnsignedBits inc = std::min(1 + inc_bits % max_inc, max_bits - twiddled_key);
+        twiddled_key += inc;
+        
+        // Generate random run length (ensure there are enough values to fill the rest).
+        size_t run_bits = 0;
+        RandomBits(run_bits);
+        size_t run_length = std::min(1 + run_bits % max_run, num_items - i);
+        if (twiddled_key == max_bits) run_length = num_items - i;
+        size_t run_end = i + run_length;
+        
+        // Fill the array.
+        UnsignedBits key = TraitsT::TwiddleOut(twiddled_key);
+        for (; i < run_end; ++i)
+        {
+            h_key_bits[i] = key;
+        }
+    }
+}
+
+
 /**
  * Initialize solution
  */
-template <bool IS_DESCENDING, typename KeyT>
+template <bool IS_DESCENDING, typename KeyT, typename OffsetT>
 void InitializeSolution(
     KeyT    *h_keys,
-    int     num_items,
+    OffsetT num_items,
     int     num_segments,
-    int     *h_segment_offsets,
+    bool    pre_sorted,
+    OffsetT *h_segment_offsets,
     int     begin_bit,
     int     end_bit,
-    int     *&h_reference_ranks,
+    OffsetT *&h_reference_ranks,
     KeyT    *&h_reference_keys)
 {
-    typedef Pair<KeyT, int> PairT;
-
-    PairT *h_pairs = new PairT[num_items];
-
-    int num_bits = end_bit - begin_bit;
-    for (int i = 0; i < num_items; ++i)
+    if (pre_sorted)
     {
-
-        // Mask off unwanted portions
-        if (num_bits < static_cast<int>(sizeof(KeyT) * 8))
+        printf("Shuffling reference solution on CPU (%d segments)\n", num_segments);
```

Comment:
The presorted branch doesn't seem to account for `num_segments`. Is this intentional?

Reply:
Yes, only a single segment is supported to simplify the implementation.

Now that I think of it, supporting multiple segments shouldn't be hard. The part that generates the keys remains the same. For permutation and computing ranks, I can perform it per-segment instead of for the whole array.

Do you want me to do it?